apps/omg/lib/omg/state/core.ex

```diff
@@ -250,75 +289,89 @@ defmodule OMG.State.Core do
-  end
+  def get_exiting_utxo_positions([%{utxo_pos: _} | _] = exit_infos, state),
+    do: exit_infos |> Enum.map(& &1.utxo_pos) |> get_exiting_utxo_positions(state)
 
   # list of full exit events (from ethereum listeners)
-  def exit_utxos([%{call_data: %{utxo_pos: _}} | _] = exit_infos, %Core{} = state) do
-    exit_infos |> Enum.map(& &1.call_data) |> exit_utxos(state)
-  end
+  def get_exiting_utxo_positions([%{call_data: %{utxo_pos: _}} | _] = exit_infos, state),
+    do: exit_infos |> Enum.map(& &1.call_data) |> get_exiting_utxo_positions(state)
 
   # list of utxo positions (encoded)
-  def exit_utxos([encoded_utxo_pos | _] = exit_infos, %Core{} = state) when is_integer(encoded_utxo_pos) do
-    exit_infos |> Enum.map(&Utxo.Position.decode!/1) |> exit_utxos(state)
-  end
+  def get_exiting_utxo_positions([encoded_utxo_pos | _] = exit_infos, %Core{}) when is_integer(encoded_utxo_pos),
+    do: Enum.map(exit_infos, &Utxo.Position.decode!/1)
 
   # list of IFE input/output piggybacked events
-  def exit_utxos([%{call_data: %{in_flight_tx: _}} | _] = in_flight_txs, %Core{} = state) do
+  def get_exiting_utxo_positions([%{call_data: %{in_flight_tx: _}} | _] = in_flight_txs, %Core{}) do
     _ = Logger.info("Recognized exits from IFE starts #{inspect(in_flight_txs)}")
 
     in_flight_txs
     |> Enum.flat_map(fn %{call_data: %{in_flight_tx: tx_bytes}} ->
       {:ok, tx} = Transaction.decode(tx_bytes)
       Transaction.get_inputs(tx)
     end)
-    |> exit_utxos(state)
   end
 
   # list of IFE input piggybacked events (they're ignored)
-  def exit_utxos([%{tx_hash: _, omg_data: %{piggyback_type: :input}} | _] = piggybacks, state) do
+  def get_exiting_utxo_positions([%{tx_hash: _, omg_data: %{piggyback_type: :input}} | _] = piggybacks, %Core{}) do
     _ = Logger.info("Ignoring input piggybacks #{inspect(piggybacks)}")
-    {:ok, {[], {[], []}}, state}
+    []
   end
 
   # list of IFE output piggybacked events. This is used by the child chain only. `OMG.Watcher.ExitProcessor` figures out
   # the utxo positions to exit on its own
-  def exit_utxos([%{tx_hash: _, omg_data: %{piggyback_type: :output}} | _] = piggybacks, state) do
+  def get_exiting_utxo_positions(
+        [%{tx_hash: _, omg_data: %{piggyback_type: :output}} | _] = piggybacks,
+        %Core{} = state
+      ) do
     _ = Logger.info("Recognized exits from piggybacks #{inspect(piggybacks)}")
 
     piggybacks
     |> Enum.map(&find_utxo_matching_piggyback(&1, state))
-    |> Enum.filter(& &1)
+    |> Enum.filter(fn utxo -> utxo != nil end)
     |> Enum.map(fn {position, _} -> position end)
-    |> exit_utxos(state)
   end
 
   # list of utxo positions (decoded)
-  def exit_utxos([Utxo.position(_, _, _) | _] = exiting_utxos, %Core{utxos: utxos} = state) do
+  def get_exiting_utxo_positions([Utxo.position(_, _, _) | _] = exiting_utxos, %Core{}), do: exiting_utxos
+
+  @doc """
+  Spends exited utxos.
+  Note: state passed here is already extended with DB.
```

Comment:
the reason why you had to leave this comment is because you introduced `Core.with_utxos/2` which makes it harder to reason about what it does with state!